src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -192,7 +192,7 @@ private[concurrent] final object Promise {
     @tailrec override final def toString: String = {
       val state = get()
       if (state.isInstanceOf[Try[T]]) "Future("+state+")"
-      else if (state.isInstanceOf[Link[T]]) state.asInstanceOf[Link[T]].promise(this).toString
+      else if (state.getClass eq classOf[Link[T]]) state.asInstanceOf[Link[T]].promise(this).toString
```

Comment:
I assume you benchmarked this and it was faster? I'm surprised the JIT can't tell that the type has no children and not walk the object's class hierarchy.

Reply:
Yeah, it was faster, and I too was surprised :(

Reply:
@mkeskells The difference was small, yes. (I run these benches: https://github.com/scala/scala/blob/76b34c455e54adea171bd618e7ea5d37cc2c4af1/test/benchmarks/src/main/scala/scala/concurrent/FutureBenchmark.scala) I found another, much bigger, source of performance gain so I am happy to switch back to using isInstanceOf (I just need to do some extra benching and cleanup first).

Reply:
it would be good to understand if and when a different approach is faster, after all if it an optimisation that the compiler can apply for us we can keep the code idiomatic and faster. 
I would ask @retronym to have a look at the when/if any optimisation is generalisable in the compiler codegen

Or if it just an optimisation to pass onto a JIT/Graal team, if it is just a runtime rather than codegen optimisation

Reply:
AFAIK, HotSpot and Graal generate equivalent code for `instanceof` and `.getClass == Foo.class` testing against an effectively final class. I'd suggest switching back to `instanceof`.

Reply:
@retronym Agreed—already done :)

Reply:
@mkeskells I'm going to compile a list of proposed changes to the compiler based on my observations, for sure!

Reply:
I don't think there's anything wrong with using `getClass` even if it's not definitely faster